Refactor reading status and reading list into enums

SergioMendolia opened the pull request:
## Proposed changes

Explain your PR here


## Checklist
- [ ] Tests are passing
- [ ] New tests have been written
- [ ] Documentation has been updated
- [ ] Translations have been updated
- [ ] Breaking changes have been avoided or documented

onigoetz commented on templates/components/InlineEditInteraction.html.twig:
```diff
@@ -1,87 +1,48 @@
-            {% set finished=true %}
-        {% endif %}
+    <div class="btn-group-vertical btn-group-sm w-100 my-2">
+    {% for case in enum('App\\Enum\\ReadStatus').cases() %}
```
I don't understand how this works.
It loops over all read statuses so it shows one button per status ?
missa no comprendo

onigoetz reviewed (commented):
The "bookinteraction" model confuses me a lot.

It's a 1:n relationship with books and that I can understand

1 person can read the book multiple times.
But the table also contains some unrelated aspects:

1. read_pages : indicates it's in progress
2. read_status : indicates it's in progress or finished
3. rating : vote on the rating of the books

This model is weird to me as the arity is strange: 

1. read_pages should be a 1:1 per user; you can't have the same book in progress multiple times
2. read_status : you can read a book multiple times, that tracks
3. rating: you should be able to vote on a book only once

Also, selecting the values is different for each use cases
1. read pages: you should be able to get only one per user
2. read status: it could be finished AND in progress at the same time. but queries only one user
3. rating: you want to retrieve the rows for all users for this one

I would recommend three tables:

1. read_log
  - book_id
  - user_id
  - date
2. read_progress
  - book_id
  - user_id
  - pages_read
3. ratings
  - book_id
  - user_id
  - rating

The way to retrieve information for each use cases becomes way clearer.

SergioMendolia commented:
@onigoetz 
it should be a 0:1 relationship per user (interaction or no interaction) but this has to translate in the code by a 0:n because there are multiple users.
- I added a unique index so one user can only have one interaction per book
- I simplified the inlineEditInteraction to make sure it works in this way
- If we want to have more than one readdate, we will indeed need a new table, but this is not there yet
- since there is only one interaction, rating and readinglist make sense in this table.
- if we ever add multiple readdates, the status will be calculated based on the new table and in that new table we will have the status of that "read" and the readpages. but again, not for now.

onigoetz commented:
> since there is only one interaction, rating and readinglist make sense in this table.

I disagree on this point, the retrieval of interactions (readinglist, pages read ...) should be done only for the current user.
Whereas the retrieval of the rating has to be done for all users. I don't expect our users to have hundreds of users on a single instance, but when showing a list of books that means loading all interactions of all books then filtering on server side.

We could even optimize the retrieval of ratings by doing it with a join and average on DB side.

Cool for the addition of an index, that will help with the bug I introduced 😬 

In the meantime could we already go forward with the refactoring to enums and discuss the addition of ratings in a separate PR? because in current state it adds them only to DB but doesn't show them anywhere

SergioMendolia commented:
> > since there is only one interaction, rating and readinglist make sense in this table.
> 
> I disagree on this point, the retrieval of interactions (readinglist, pages read ...) should be done only for the current user. Whereas the retrieval of the rating has to be done for all users. I don't expect our users to have hundreds of users on a single instance, but when showing a list of books that means loading all interactions of all books then filtering on server side.
> 
> We could even optimize the retrieval of ratings by doing it with a join and average on DB side.
> 
> Cool for the addition of an index, that will help with the bug I introduced 😬
> 
> In the meantime could we already go forward with the refactoring to enums and discuss the addition of ratings in a separate PR? because in current state it adds them only to DB but doesn't show them anywhere

ratings are only personal for now, not loading other user's ratings

onigoetz commented:
OOOH, I didn't get that, my bad :D

onigoetz commented on templates/book/_interaction.html.twig:
```diff
@@ -1,12 +1,2 @@
-    {{ component('InlineEditInteraction',{'user':app.user,'book':book, 'shelves': shelves, 'buttonWithText': true}) }}
-{% endif %}
\ No newline at end of file
+{{ component('InlineEditInteraction',{'book':book, 'buttonWithText': true}) }}
```
Do we need to keep the file if the only thing it does is include the component?

onigoetz commented on src/Controller/BookController.php:
```diff
@@ -199,15 +200,18 @@ public function read(
+        if ($interaction->getReadStatus() !== ReadStatus::Finished && $page === $book->getPageNumber()) {
+            $interaction->setReadStatus(ReadStatus::Finished);
+
+            // TODO Add next unread in serie to reading list?
```
Could be nice indeed, but let's already merge this 😄

onigoetz reviewed (approved):
<img width="260" alt="image" src="https://github.com/user-attachments/assets/5575d4d4-3451-4e10-9a67-5e264dc24285" />

I'm not a big fan of how the menu looks like. But it is functional so let's merge and discuss next steps after 😄 

I also added a few comments, but nothing that's problematic
